Cache code generation during triton template expansion and enable it  for mm_template.

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151773

In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes.

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part.
This is expected to save 20% of the model I am working on.

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. !
To handle those we use a record and replay approach, where we record the functions that are called during
code generation that effect those outputs and replay them at a cache hit.

Effect on the current benchmark on a local run on dev server.
mm_loop. 24115830838 -> 18362098019
mm_loop_dynamic 30506097176-> 25697270062

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

laithsakka committed:
......

[ghstack-poisoned]

laithsakka committed:
Update on "Cache code generation during triton template expansion for mm_template."


In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes.

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part.
This is expected to save 20% of the model I am working on.

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. !
To handle those we use a record and replay approach, where we record the functions that are called during
code generation that effect those outputs and replay them at a cache hit.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
address comments

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -367,6 +371,49 @@ def __init__(
+    def input_dependent_preserved_state(self) -> str:
+        return repr(
+            [
+                self.args.input_buffers,
+                self.args.sizevars,
+                self.args.workspace_args,
```
Can we just include all of `self.args` ? the initial pr for this missed `self.args.workspace_args` and we didnt have anyway of knowing.

laithsakka replied:
as i mentioned in the comment on the previous PR adding all of  self.args have a bad performance effect, I did try that. Namely adding  self.output_buffers is not good here

eellison replied:
Okay, for the debug mode, can we add an assertions along these lines?

laithsakka replied:
can you elaborate more on that type of assertions you want me to add here?

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1068,19 +1121,138 @@ def _jinja2_env():
+            has_flexiable_layout()
+            or subgraphs is not None
+            or workspace_arg is not None
+            or epilogue_fn is not identity
```
the epilogue_fn seems easy to support.. but can leave for follow up

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1068,19 +1121,138 @@ def _jinja2_env():
+                    return True
+            return False
+
+        # we do not cache under those conditions right now.
```
When do we ever have flexible layouts?

laithsakka replied:
I thought you asked me in the previous PR to assert that we do not pass flexible_layout. 
or did you mean just an assert?

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1068,19 +1121,138 @@ def _jinja2_env():
+                ]
+            )
+
+        def has_flexiable_layout() -> bool:
```
nit: `has_flexiable_layout` -> `has_flexible_layout`

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1201,19 +1389,51 @@ def make_extra() -> str:
         assert code is not None and extra is not None
 
+        if cache_hit and self.test_cache:
+            with (
```
nit: factor this out to helper function, or as optoinal hook to avoid putting the test logic inline in main logic

laithsakka replied:
done

eellison reviewed (commented):
Looks good - but could we more explicitly disable this with epilogue fusion ? not sure if that is being done currently

laithsakka commented:
> Looks good - but could we more explicitly disable this with epilogue fusion ? not sure if that is being done currently

here is my response to that in the original PR in case you missed
<img width="893" alt="Screenshot 2025-05-01 at 9 23 09 PM" src="https://github.com/user-attachments/assets/2855cd6a-76d2-44e4-a32c-338541665fb7" />

I am not sure what else we can since that happen

eellison commented:
Yes, les explicitly disable caching in the epilogue case.

laithsakka committed:
Update on "Cache code generation during triton template expansion and enable it  for mm_template."


In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes.

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part.
This is expected to save 20% of the model I am working on.

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. !
To handle those we use a record and replay approach, where we record the functions that are called during
code generation that effect those outputs and replay them at a cache hit.

Effect on the current benchmark on a local run on dev server.
mm_loop. 24115830838 -> 18362098019
mm_loop_dynamic 30506097176-> 25697270062

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
Address comments,  also 
1) I enabled testing cache by default with debug mode is on. 
2) added a global flag to turn on caching will make opt in by ads model for now. 
3) only do the caching throw call to append_choice_caller  tritonTemplate to exclude others paths. 
4) refactor testing into its own function

laithsakka commented:
@eellison  can you take another look

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1079,6 +1134,115 @@ class GenerateAndLoadResult(NamedTuple):
+
+        return repr(
+            {
+                "input_nodes": [
+                    layout_key(input.get_layout()) for input in input_nodes
+                ],
+                "num_stages": num_stages,
+                "num_warps": num_warps,
+                "prefix_args": prefix_args,
+                "suffix_args": suffix_args,
+                "call_sizes": call_sizes,
+                "layout": layout_key(layout),
+                "num_consumer_groups": num_consumer_groups,
+                "num_buffers_warp_spec": num_buffers_warp_spec,
+                "kwargs": kwargs,
```
See: https://github.com/pytorch/pytorch/pull/148437/files#diff-1f0410c70fcfc019bc1de443fdaa8153947c374646b7c7ad25e75427902a5397R657-R658

We're skipping `waves_per_eu` and `kpack`

laithsakka replied:
Those options are not passed to the "generate" call of TritonTemplate(KernelTemplate), hence they are not relevant. 
but lets say someone added some args   TritonTemplate(KernelTemplate). the have to pass them to generate_and_laod at that point the code is very clearly calling the cache the first line 
I do not think any user will miss those, if those are part of kwargs they will be auto detected also.
so i dont see the issue tbh. 


<img width="705" alt="Screenshot 2025-05-12 at 5 36 16 PM" src="https://github.com/user-attachments/assets/c4325aa9-48f4-4544-abfb-1893d2599244" />

laithsakka replied:
those are included in the cache key automatically since they are part of kwargs.

In general anyone adding a configs its either:
1) part of kwargs then it will be included automatically. 
2) a kernel option in such case they have ti pass it to generate_and_load since kernel optins created there, 
if they add a new argument to  generate_and_load caching path will be clear i even added a comment
```
    # NOTE: MAKE SURE THAT ANY ARGUMENT ADDED TO THIS FUNCTION IS PROPERLY HANDLED IN _generated_code_cache.make_key.
    def generate_and_load(
        self,
        input_nodes: tuple[ir.IRNode],
        num_stages: int,
        num_warps: int,
        call_sizes: list[sympy.core.symbol.Symbol],
        prefix_args: int,
        suffix_args: int,
        epilogue_fn: Optional[Callable[..., Any]],
        subgraphs: Optional[list[ir.Buffer]],
        workspace_arg: Optional[WorkspaceArg],
        num_consumer_groups: int,
        num_buffers_warp_spec: int,
        layout: ir.Layout,
        kwargs: dict[str, Any],
        generate_with_caching,
    ) -> Optional[GenerateAndLoadResult]:
        """Generate the python code and load it into the current process"""
        caching_enabled = (
            generate_with_caching
            and torch._inductor.config.enable_caching_generated_triton_templates
        )

        cache_key = None
        if caching_enabled:
            cache_key = self._generated_code_cache.make_key(
                input_nodes,
                num_stages,
                num_warps,
                call_sizes,
                prefix_args,
                suffix_args,
                epilogue_fn,
                subgraphs,
                workspace_arg,
                layout,
                num_consumer_groups,
                num_buffers_warp_spec,
                kwargs,
            )

```

eellison reviewed (commented):
Someone recently added more options but we are not using them on cache key. 

We need a better story for discoverability of this, or this will very quickly get incorrect. We should infer the fields more automatically or there should be an error

laithsakka committed:
Update on "Cache code generation during triton template expansion and enable it  for mm_template."


In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes.

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part.
This is expected to save 20% of the model I am working on.

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. !
To handle those we use a record and replay approach, where we record the functions that are called during
code generation that effect those outputs and replay them at a cache hit.

Effect on the current benchmark on a local run on dev server.
mm_loop. 24115830838 -> 18362098019
mm_loop_dynamic 30506097176-> 25697270062

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
added a comment
    # NOTE: MAKE SURE THAT ANY ARGUMENT ADDED TO THIS FUNCTION IS PROPERLY HANDLED IN _generated_code_cache.make_key.

eellison reviewed (commented):
we discussed adding one more test, then should be good to go.

laithsakka committed:
Update on "Cache code generation during triton template expansion and enable it  for mm_template."


In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes.

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part.
This is expected to save 20% of the model I am working on.

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. !
To handle those we use a record and replay approach, where we record the functions that are called during
code generation that effect those outputs and replay them at a cache hit.

Effect on the current benchmark on a local run on dev server.
mm_loop. 24115830838 -> 18362098019
mm_loop_dynamic 30506097176-> 25697270062

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Cache code generation during triton template expansion and enable it  for mm_template."


In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes.

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part.
This is expected to save 20% of the model I am working on.

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. !
To handle those we use a record and replay approach, where we record the functions that are called during
code generation that effect those outputs and replay them at a cache hit.

Effect on the current benchmark on a local run on dev server.
mm_loop. 24115830838 -> 18362098019
mm_loop_dynamic 30506097176-> 25697270062

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge -i

laithsakka committed:
Update on "Cache code generation during triton template expansion and enable it  for mm_template."


In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes.

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part.
This is expected to save 20% of the model I am working on.

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. !
To handle those we use a record and replay approach, where we record the functions that are called during
code generation that effect those outputs and replay them at a cache hit.

Effect on the current benchmark on a local run on dev server.
mm_loop. 24115830838 -> 18362098019
mm_loop_dynamic 30506097176-> 25697270062

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Cache code generation during triton template expansion and enable it  for mm_template."


In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes.

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part.
This is expected to save 20% of the model I am working on.

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. !
To handle those we use a record and replay approach, where we record the functions that are called during
code generation that effect those outputs and replay them at a cache hit.

Effect on the current benchmark on a local run on dev server.
mm_loop. 24115830838 -> 18362098019
mm_loop_dynamic 30506097176-> 25697270062

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge

malfet commented:
@pytorchbot revert -m "It broke ROCm, see https://hud.pytorch.org/hud/pytorch/pytorch/f9aa3bae8cec28528de819863b04eae75089c4c5/1?per_page=50&name_filter=rocm%20%2F%20linux-jammy&mergeEphemeralLF=true" -c nosignal

laithsakka committed:
Update on "Cache code generation during triton template expansion and enable it  for mm_template."


In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes.

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part.
This is expected to save 20% of the model I am working on.

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. !
To handle those we use a record and replay approach, where we record the functions that are called during
code generation that effect those outputs and replay them at a cache hit.

Effect on the current benchmark on a local run on dev server.
mm_loop. 24115830838 -> 18362098019
mm_loop_dynamic 30506097176-> 25697270062

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Cache code generation during triton template expansion and enable it  for mm_template."


In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes.

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part.
This is expected to save 20% of the model I am working on.

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. !
To handle those we use a record and replay approach, where we record the functions that are called during
code generation that effect those outputs and replay them at a cache hit.

Effect on the current benchmark on a local run on dev server.
mm_loop. 24115830838 -> 18362098019
mm_loop_dynamic 30506097176-> 25697270062

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge -f "just rebaed"

yangw-dev commented:
@pytorchbot revert -m "sorry, it seems your pr breaks the test internally 
D75169759" -c ghfirst

yangw-dev commented:
@pytorchbot revert -m "sorry, it seems your pr breaks the test internally D75169759" -c ghfirst

yangw-dev commented:
@pytorchbot revert -m "sorry, it seems your pr breaks the test internally D75169759" -c ghfirst


FAIL: test_triton_template_generated_code_caching (caffe2.test.inductor.test_max_autotune.TestMaxAutotune)
----------------------------------------------------------------------
    raise error_metas.pop()[0].to_error(  # type: ignore[index]
AssertionError: String comparison failed: '{\'i[406 chars]32\':False,\'USE_FAST_ACCUM\':False,\'ACC_TYPE[74 chars]:8}}' != '{\'i[406 chars]32\':True,\'USE_FAST_ACCUM\':False,\'ACC_TYPE\[73 chars]:8}}'
Diff is 1769 characters long. Set self.maxDiff to None to see it.

----------------------------------------------------------------------
Ran 1 test in 3.846s

FAILED (failures=1)
AssertionError: String comparison failed: '{\'i[406 chars]32\':False,\'USE_FAST_ACCUM\':False,\'ACC_TYPE[74 chars]:8}}' != '{\'i[406 chars]32\':True,\'USE_FAST_ACCUM\':False,\'ACC_TYPE\[73 chars]:8}}'
